{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "331086f6-954a-4e14-a50e-04d6005ee889",
   "metadata": {
    "tags": []
   },
   "source": [
    "<div class=\"jupyter-biolm-header\">\n",
    "    <img style=\"float: left; padding-right: 10px; height: 60px\" src=\"https://d31e6ufxekikrt.cloudfront.net/static/ui/images/logo.png\">\n",
    "    <p>\n",
    "    <br>\n",
    "    <br>\n",
    "    <br>\n",
    "    </p>\n",
    "</div>\n",
    "\n",
    "## ProGen2-OAS Antibody Sequence and Structure Generation Tutorial (Utomilumab)\n",
    "\n",
    "In this tutorial, we demonstrates the use of the BioLM.ai API with the ProGen2-OAS model to generate a library and predict the structure of Utomilumab antibody sequences that are annotated and numbered, have Levenshtein distances calculated, and with new CDR3 region sequences. \n",
    "\n",
    "\n",
    "<br>\n",
    "\n",
    "<table class=\"jupyter-biolm-header-table\" style=\"width: 100%; border-collapse: collapse; background-color: white; float: left;\">\n",
    "    <tr>\n",
    "        <td style=\"text-align: left; vertical-align: middle; background-color: white;\">\n",
    "            <img src=\"https://www.svgrepo.com/show/354202/postman-icon.svg\" style=\"height: 15px; float: left; padding-right: 10px;\"><a href=\"https://api.biolm.ai/#2424fa3a-144c-4d8e-8391-9e69b9df15d0\">  <h5 style=\"margin: 0;\"><b>Postman API Docs</b></h5></a>\n",
    "        </td>\n",
    "        <td style=\"text-align: left; vertical-align: middle; background-color: white;\">\n",
    "            <img src=\"https://upload.wikimedia.org/wikipedia/commons/thumb/c/c3/Python-logo-notext.svg/1869px-Python-logo-notext.svg.png\" style=\"height: 15px; float: left; padding-right: 10px;\"><a href=\"https://docs.biolm.ai/en/latest/model-docs/progen2/ProGen2_API.html\"><h5 style=\"margin: 0;\"><b>Python SDK Docs</b></h5></a>\n",
    "        </td>\n",
    "        <td style=\"text-align: left; vertical-align: middle; background-color: white;\">\n",
    "            <img src=\"https://github.githubassets.com/assets/GitHub-Mark-ea2971cee799.png\" style=\"height: 15px; float: left; padding-right: 10px;\"><a href=\"https://github.com/salesforce/progen/tree/main/progen2\"><h5 style=\"margin: 0;\"><b>Github Link</b></h5></a>\n",
    "        </td>\n",
    "        <td style=\"text-align: left; vertical-align: middle; background-color: white;\">\n",
    "            <img src=\"https://info.arxiv.org/brand/images/brand-logomark-primary-large.jpg\" style=\"height: 15px; float: left; padding-right: 10px;\"><a href=\"https://arxiv.org/pdf/2206.13517.pdf\"><h5 style=\"margin: 0;\"><b>Paper Link</b></h5></a>\n",
    "        </td>\n",
    "    </tr>\n",
    "</table>\n",
    "\n",
    "<br>\n",
    "\n",
    "---"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f88e9860-4315-4534-b1bd-f557a0e15f43",
   "metadata": {},
   "source": [
    "# Utomilumab: Our Antibody of Interest.\n",
    "Utomilumab is a is a fully human IgG2 monoclonal antibody that targets the CD137 receptor (also known as 4-1BB), which is a member of the tumor necrosis factor (TNF) receptor family. CD137 is a co-stimulatory molecule expressed on the surface of activated T cells and other immune cells. The engagement of CD137 with its ligand or agonistic antibodies like utomilumab can enhance T cell proliferation, cytokine production, and survival, leading to an amplified immune response. You can learn more about its mechanism of action \n",
    "<u>[here](https://cdn.pfizer.com/pfizercom/news/asco/Utomilumab_Fact_Sheet_23MAY2018.PDF)<u>\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "019309cd-3b66-47e4-8dbf-9eee29230f92",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Set your API Token\n",
    "\n",
    "To use the BioLM API, you need an API token. You can get one from the <u>[User API Tokens](https://biolm.ai/ui/accounts/user-api-tokens/)<u> page.\n",
    "\n",
    "Paste the API token in the cell below, as the value of the variable BIOLMAI_TOKEN."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "72696982-c675-4cb2-b987-29431a493c20",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "BIOLMAI_TOKEN = \" \" # !!! YOUR API TOKEN HERE !!!"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d8430165-4ce9-495c-9e88-35f0bf11b992",
   "metadata": {},
   "source": [
    "## API Call with Python Requests"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "6733f0a8-96e7-4bce-a3fc-a62c36a81c3a",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "#The below is needed if running this notebook in a Pyodide environment:\n",
    "\n",
    "try:\n",
    "    # Install packages to make API requests in JuputerLite\n",
    "    import micropip\n",
    "    await micropip.install('requests')\n",
    "    await micropip.install('pandas')\n",
    "    await micropip.install('numpy')\n",
    "    await micropip.install('pyodide-http')\n",
    "    # Patch requests for in-browser support\n",
    "    import pyodide_http\n",
    "    pyodide_http.patch_all()\n",
    "except ModuleNotFoundError:\n",
    "    pass  # Won't be using micropip outside of JLite\n",
    "    \n",
    "import requests\n",
    "from IPython.display import JSON  # Helpful UI for JSON display\n",
    "\n",
    "# Note: sadie-antibody and Levenshtein will not install in a Pyodide environment. They can be installed in a local Jupyter notebook or terminal using \"pip install\". \n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "16bd6271-d081-4be6-98b3-e91e0118b215",
   "metadata": {},
   "source": [
    "# Import Necessary Packages and Dependencies\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "971ee18d-b965-420f-89d2-efa6aabcccd0",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "import warnings\n",
    "warnings.filterwarnings('ignore', category=FutureWarning, module='')\n",
    "\n",
    "import pandas as pd\n",
    "import requests\n",
    "import json\n",
    "import zipfile\n",
    "import Levenshtein as lev\n",
    "import os\n",
    "from sadie.renumbering import Renumbering\n",
    "from sadie.renumbering.result import NumberingResults\n",
    "from IPython.display import JSON "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1cfe6d08-e930-42e9-84e0-37ba44a69356",
   "metadata": {},
   "source": [
    "## Initializing Constants, File Paths, and Sequence Data\n",
    "\n",
    "The Utomilumab sequnce can be obtained from <u>[Thera-SAbDab](https://opig.stats.ox.ac.uk/webapps/sabdab-sabpred/therasabdab/therasummary/?INN=utomilumab&format=Whole+mAb&clintrial=Phase-II&status=Active&target=TNFRSF9/CD137/4-1BB&yearprop=2016&isotype=G2&heavy1=EVQLVQSGAEVKKPGESLRISCKGSGYSFSTYWISWVRQMPGKGLEWMGKIYPGDSYTNYSPSFQGQVTISADKSISTAYLQWSSLKASDTAMYYCARGYGIFDYWGQGTLVTVSS&light1=SYELTQPPSVSVSPGQTASITCSGDNIGDQYAHWYQQKPGQSPVLVIYQDKNRPSGIPERFSGSNSGNTATLTISGTQAMDEADYYCATYTGFGSLAVFGGGTKLTVL&heavy2=na&light2=na&struc100=6a3w:GH:DE:AB:JK/6mi2:DE:AB&struc99=None&struc95to98=None&yearrec=2017&companies=Dana-Farber+Cancer+Institute;Immatics+US;Kite+Pharma;Kyowa+Hakko+Kirin;M.+D.+Anderson+Cancer+Center;Pfizer&cond_approved=na&cond_active=Breast+cancer;Oropharyngeal+cancer;Solid+tumours;B-cell+lymphoma;Diffuse+large+B+cell+lymphoma;Follicular+lymphoma;Ovarian+cancer&cond_disc=Non-Hodgkin%27s+lymphoma;Colorectal+cancer&dev_tech=MorphoSys+HuCAL+Phage+Display&notes=)<u>\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "55c2f011-df34-497a-978f-6ea7b7c6cad3",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Move on to Functions 1 and 2\n"
     ]
    }
   ],
   "source": [
    "# Constants\n",
    "json_result_file = 'ProGen2_OAS_sequences.json'\n",
    "csv_file_path = 'ProGen2_OAS_sequences.csv'\n",
    "json_directory = \"json_results\"\n",
    "zip_file_path = \"json_results.zip\"\n",
    "\n",
    "# The process begins by including the wild-type (wt) heavy chain sequence and a truncated version up to (and including) the fwr3 region.\n",
    "\n",
    "wt_sequence = \"EVQLVQSGAEVKKPGESLRISCKGSGYSFSTYWISWVRQMPGKGLEWMGKIYPGDSYTNYSPSFQGQVTISADKSISTAYLQWSSLKASDTAMYYCARGYGIFDYWGQGTLVTVSS\" #Wild-type (wt) heavy chain sequence. length = 116 amino acids. \n",
    "wt_until_fwr3 = \"EVQLVQSGAEVKKPGESLRISCKGSGYSFSTYWISWVRQMPGKGLEWMGKIYPGDSYTNYSPSFQGQVTISADKSISTAYLQWSSL\" #Wild-type (wt) heavy chain sequence up to and uncluding fwr3. length= 86 amino acids (can be customized in Function 1).\n",
    "\n",
    "print(\"Move on to Functions 1 and 2\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d28a718b-eb75-4cb2-ae17-b22d38f4d441",
   "metadata": {},
   "source": [
    "## Function 1 and Function 2\n",
    "\n",
    "The below few code blocks are going to run Function 1 and Function 2. Once these two code blocks are run, the user can call both functions a few code blocks below under the title \"Calling Function 1 and 2\", to generate the sequence csv file that has been annotated and numbered, with Levenshtein distance calculated. "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d5dc67b3-2752-4855-a50b-87f05dcb93c2",
   "metadata": {},
   "source": [
    "## Function 1: Generating Sequences using ProGen2-OAS via the BioLM API\n",
    "\n",
    "Sequence Generation: The \"generate_progen2_sequences\" function below sends requests to the ProGen2-OAS API to generate new antibody sequences based on the provided wild-type sequence. The function allows for various parameters such as num_requests (to control the #API requests sent), temperature, max_length, and top_p to control the generation process.\n",
    "For more information on these parameters, refer the BioLM.ai <u>[API docs](https://docs.biolm.ai/en/latest/model-docs/progen2/ProGen2_API.html)<u>\n",
    "\n",
    "The responses are saved as JSON files. The resulting sequences and their JSON files are then zipped for storage.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "45209457-b5dd-4118-b51e-07ec40555944",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Function 1:\n",
    "\n",
    "def generate_progen2_sequences(wt_sequence, truncate_length=None, t=0.8, top_p=0.6, max_length=135, num_requests=2, sequences_per_request=3):  \n",
    "    url = \"https://biolm.ai/api/v2/progen2-oas/generate/\"\n",
    "    headers = {'Content-Type': 'application/json', \"Authorization\": f\"Token {BIOLMAI_TOKEN.strip()}\"}\n",
    "    os.makedirs(json_directory, exist_ok=True)\n",
    "\n",
    "    if truncate_length is not None:\n",
    "        wt_until_fwr3 = wt_sequence[:truncate_length]\n",
    "\n",
    "    all_sequences = []             # Initializing the generated sequences list.\n",
    "    for i in range(num_requests):  # Iterating over the number of API requests to make.\n",
    "        payload = json.dumps({     # Preparing the payload for the POST request as a JSON string\n",
    "            \"params\": {\"temperature\": t, \"top_p\": top_p, \"num_samples\": sequences_per_request, \"max_length\": max_length},\n",
    "            \"items\": [{\"context\": wt_until_fwr3}]\n",
    "        })\n",
    "        response = requests.post(url, headers=headers, data=payload) #response to the POST request is stored in a variable called 'response'. Note, 'response' is an object here. \n",
    "\n",
    "        if response.status_code == 200: #If response = 200, a successful request has been made.\n",
    "            data = response.json()\n",
    "            json_file_path = os.path.join(json_directory, f'json_result_{i + 1}.json')\n",
    "            with open(json_file_path, 'w') as f:\n",
    "                json.dump(data, f, indent=4)\n",
    "            print(f\"JSON results saved to {json_file_path}\")\n",
    "\n",
    "             \n",
    "            # The 'results' is ackey in a dictionary. That dictionary contains a list and another list within that list.\n",
    "            # The inner-most list contains three dictionaries, whose keys are 'll_sum','ll_mean' and 'sequence', and each key contains associated values.\n",
    "            # We will filter sequences with 'll_mean' greater than -2 and return them. \n",
    "            filtered_sequences = filtered_sequences = [(item[\"sequence\"], item[\"ll_mean\"]) for item in data[\"results\"][0] if item['ll_mean'] > -2]  \n",
    "\n",
    "            all_sequences.extend(filtered_sequences)\n",
    "        else:\n",
    "            print(f\"Error during generation: {response.text}\")\n",
    "    \n",
    "    with zipfile.ZipFile(zip_file_path, 'w') as zipf:  # Creating a ZIP file to compress all the JSON files containing the results from the API requests.\n",
    "        for root, _, files in os.walk(json_directory):\n",
    "            for file in files:\n",
    "                zipf.write(os.path.join(root, file), arcname=file)\n",
    "    print(f\"JSON results are zipped in {zip_file_path}\")\n",
    "    print (\"Completed Function 1\")\n",
    "\n",
    "    return all_sequences # We now have a list of ProGen2-OAS generated sequences generated via the BioLM.ai API that have been filtered based on the 'll_mean' value\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "49bebe7f-fc89-4dc1-a60e-5373824aeb25",
   "metadata": {},
   "source": [
    "## Function 2: Sequence Annotation and Distance Calculation\n",
    "\n",
    "The \"annotate_and_calculate_distances\" function below uses the SADIE renumbering API to annotate both the wild-type sequence and the generated sequences. The below function also calculates the Levenshtein distance between each generated sequence and the wild-type sequence to assess their similarity. The results, including the sequences, their annotations, and distances, are saved in a CSV file for further analysis (such as structure prediction). \n",
    "\n",
    "<i> You can learn more about Sadie-Antibody (Sequencing Analysis and Data library for Immunoinformatics Exploration) <u>[here](https://pypi.org/project/sadie-antibody/)<u><i>\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "b0aba085-0922-4864-aabd-3b0fd8f5d6e3",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Function 2:\n",
    "\n",
    "def annotate_and_calculate_distances(sequences):\n",
    "    renumbering_api = Renumbering(scheme=\"imgt\", region_assign=\"imgt\", run_multiproc=True)  # creating an instance of the Renumbering class from the SADIE-Antibody library. We use the default setting of \"imgt\" numbering scheme and the \"imgt\" region assignment.\n",
    "    results = []  # Initializing empty list for results. Note: we are using this list to collect dictionaries.\n",
    "\n",
    "    wt_numbering_result = renumbering_api.run_single(\"wt_sequence\", wt_sequence)\n",
    "    wt_result_dict = {key: value[0] if isinstance(value, dict) else value for key, value in wt_numbering_result.to_dict().items()}  # Flattening the dictionary. By flattening any nested dictionaries, the values are more straightforward to work with when creating a CSV file. Note: a dictionary is relevant if we are modifying the result or adding additional information to it.\n",
    "    wt_result_dict[\"Levenshtein_distance\"] = None  # Indicating no distance calculation needed for Wild-type sequence.\n",
    "    wt_result_dict['Id'] = 'wt_sequence'           # Label the row as \"wt_sequence\" (ie, row for the resulting csv file).\n",
    "    results.append(wt_result_dict)                 # Adding this dictionary to the results list.  \n",
    "\n",
    "    for i, (sequence, ll_mean_value) in enumerate(sequences):                                   #Iterating over the sequence list. \n",
    "        numbering_result = renumbering_api.run_single(f\"generated_sequence_{i + 1}\", sequence)  # 'run_single' method of the renumbering API being applied, Use of 'i+1' and f-string to make the generated sequence list more 'human-readable': ie, 'row will be named 'generated_sequence_1' instead of generated_sequence_0'\n",
    "        result_dict = {key: value[0] if isinstance(value, dict) else value for key, value in numbering_result.to_dict().items()} # Flatten the dictionary. Here, we are iterating over each key-value pair in this dictionary. The code takes the first element of a value if it is a dictionary (to flatten it) or keeps the value unchanged if it is not a dictionary.\n",
    "        result_dict[\"Levenshtein_distance\"] = lev.distance(wt_sequence, sequence) \n",
    "        result_dict['Id'] = f'generated_sequence_{i + 1}' # Label the row with the sequence number. Ie, 'generated_sequence_1', 'generated_sequence_2 'and so on...\n",
    "        result_dict[\"ll_mean\"] = ll_mean_value\n",
    "        results.append(result_dict) \n",
    "\n",
    "    results_df = pd.DataFrame(results)  # Creating a DataFrame from the results list. Each dictionary in the list becomes a row in the DataFrame.\n",
    "    results_df.to_csv(csv_file_path, index=False)\n",
    "    print(f\"Results saved to {csv_file_path}\")\n",
    "    print (\"Completed Function 2\")\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2809db6b-2172-4bb2-b812-b776f63695df",
   "metadata": {},
   "source": [
    "# Calling Function 1 and 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "66772db1-6a04-4baf-829b-9d756d73bf80",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "JSON results saved to json_results/json_result_1.json\n",
      "JSON results saved to json_results/json_result_2.json\n",
      "JSON results are zipped in json_results.zip\n",
      "Completed Function 1\n",
      "Results saved to ProGen2_OAS_sequences.csv\n",
      "Completed Function 2\n"
     ]
    }
   ],
   "source": [
    "generated_sequences = generate_progen2_sequences(wt_sequence, truncate_length=86) \n",
    "if generated_sequences:\n",
    "    annotate_and_calculate_distances(generated_sequences)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b0d8a34c-b2d4-4fcc-95db-2592383d7f93",
   "metadata": {},
   "source": [
    "## Examing the Generated CSV File"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "31da36dc-1f35-4f3f-beee-86b00d2bb181",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Id</th>\n",
       "      <th>sequence</th>\n",
       "      <th>domain_no</th>\n",
       "      <th>hmm_species</th>\n",
       "      <th>chain_type</th>\n",
       "      <th>e-value</th>\n",
       "      <th>score</th>\n",
       "      <th>seqstart_index</th>\n",
       "      <th>seqend_index</th>\n",
       "      <th>identity_species</th>\n",
       "      <th>...</th>\n",
       "      <th>fwr3_aa_gaps</th>\n",
       "      <th>fwr3_aa_no_gaps</th>\n",
       "      <th>cdr3_aa_gaps</th>\n",
       "      <th>cdr3_aa_no_gaps</th>\n",
       "      <th>fwr4_aa_gaps</th>\n",
       "      <th>fwr4_aa_no_gaps</th>\n",
       "      <th>leader</th>\n",
       "      <th>follow</th>\n",
       "      <th>Levenshtein_distance</th>\n",
       "      <th>ll_mean</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>wt_sequence</td>\n",
       "      <td>EVQLVQSGAEVKKPGESLRISCKGSGYSFSTYWISWVRQMPGKGLE...</td>\n",
       "      <td>0</td>\n",
       "      <td>human</td>\n",
       "      <td>H</td>\n",
       "      <td>0.0</td>\n",
       "      <td>185.000</td>\n",
       "      <td>0</td>\n",
       "      <td>115</td>\n",
       "      <td>human</td>\n",
       "      <td>...</td>\n",
       "      <td>NYSPSFQ-GQVTISADKSISTAYLQWSSLKASDTAMYYC</td>\n",
       "      <td>NYSPSFQGQVTISADKSISTAYLQWSSLKASDTAMYYC</td>\n",
       "      <td>ARGYG----IFDY</td>\n",
       "      <td>ARGYGIFDY</td>\n",
       "      <td>WGQGTLVTVSS</td>\n",
       "      <td>WGQGTLVTVSS</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>generated_sequence_1</td>\n",
       "      <td>EVQLVQSGAEVKKPGESLRISCKGSGYSFSTYWISWVRQMPGKGLE...</td>\n",
       "      <td>0</td>\n",
       "      <td>human</td>\n",
       "      <td>H</td>\n",
       "      <td>0.0</td>\n",
       "      <td>183.375</td>\n",
       "      <td>0</td>\n",
       "      <td>117</td>\n",
       "      <td>human</td>\n",
       "      <td>...</td>\n",
       "      <td>NYSPSFQ-GQVTISADKSISTAYLQWSSLKASDTAMYYC</td>\n",
       "      <td>NYSPSFQGQVTISADKSISTAYLQWSSLKASDTAMYYC</td>\n",
       "      <td>ARRGYS--YGYDY</td>\n",
       "      <td>ARRGYSYGYDY</td>\n",
       "      <td>WGQGTLVTVSS</td>\n",
       "      <td>WGQGTLVTVSS</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5.0</td>\n",
       "      <td>-0.411147</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>generated_sequence_2</td>\n",
       "      <td>EVQLVQSGAEVKKPGESLRISCKGSGYSFSTYWISWVRQMPGKGLE...</td>\n",
       "      <td>0</td>\n",
       "      <td>human</td>\n",
       "      <td>H</td>\n",
       "      <td>0.0</td>\n",
       "      <td>169.500</td>\n",
       "      <td>0</td>\n",
       "      <td>131</td>\n",
       "      <td>human</td>\n",
       "      <td>...</td>\n",
       "      <td>NYSPSFQ-GQVTISADKSISTAYLQWSSLKASDTAMYYC</td>\n",
       "      <td>NYSPSFQGQVTISADKSISTAYLQWSSLKASDTAMYYC</td>\n",
       "      <td>ARHSPRGGYYDSSGYYPYYYYGMDV</td>\n",
       "      <td>ARHSPRGGYYDSSGYYPYYYYGMDV</td>\n",
       "      <td>WGQGTTVTVSS</td>\n",
       "      <td>WGQGTTVTVSS</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>20.0</td>\n",
       "      <td>-0.479769</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>generated_sequence_3</td>\n",
       "      <td>EVQLVQSGAEVKKPGESLRISCKGSGYSFSTYWISWVRQMPGKGLE...</td>\n",
       "      <td>0</td>\n",
       "      <td>human</td>\n",
       "      <td>H</td>\n",
       "      <td>0.0</td>\n",
       "      <td>183.250</td>\n",
       "      <td>0</td>\n",
       "      <td>123</td>\n",
       "      <td>human</td>\n",
       "      <td>...</td>\n",
       "      <td>NYSPSFQ-GQVTISADKSISTAYLQWSSLKASDTAMYYC</td>\n",
       "      <td>NYSPSFQGQVTISADKSISTAYLQWSSLKASDTAMYYC</td>\n",
       "      <td>ARHGDYYDSSGYYPFDY</td>\n",
       "      <td>ARHGDYYDSSGYYPFDY</td>\n",
       "      <td>WGQGTLVTVSS</td>\n",
       "      <td>WGQGTLVTVSS</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>9.0</td>\n",
       "      <td>-0.419964</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>generated_sequence_4</td>\n",
       "      <td>EVQLVQSGAEVKKPGESLRISCKGSGYSFSTYWISWVRQMPGKGLE...</td>\n",
       "      <td>0</td>\n",
       "      <td>human</td>\n",
       "      <td>H</td>\n",
       "      <td>0.0</td>\n",
       "      <td>182.500</td>\n",
       "      <td>0</td>\n",
       "      <td>122</td>\n",
       "      <td>human</td>\n",
       "      <td>...</td>\n",
       "      <td>NYSPSFQ-GQVTISADKSISTAYLQWSSLKASDTAMYYC</td>\n",
       "      <td>NYSPSFQGQVTISADKSISTAYLQWSSLKASDTAMYYC</td>\n",
       "      <td>ARRGAGGTYYYYGMDV</td>\n",
       "      <td>ARRGAGGTYYYYGMDV</td>\n",
       "      <td>WGQGTTVTVSS</td>\n",
       "      <td>WGQGTTVTVSS</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>12.0</td>\n",
       "      <td>-0.452040</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>generated_sequence_5</td>\n",
       "      <td>EVQLVQSGAEVKKPGESLRISCKGSGYSFSTYWISWVRQMPGKGLE...</td>\n",
       "      <td>0</td>\n",
       "      <td>human</td>\n",
       "      <td>H</td>\n",
       "      <td>0.0</td>\n",
       "      <td>167.750</td>\n",
       "      <td>0</td>\n",
       "      <td>131</td>\n",
       "      <td>human</td>\n",
       "      <td>...</td>\n",
       "      <td>NYSPSFQ-GQVTISADKSISTAYLQWSSLKASDTAMYYC</td>\n",
       "      <td>NYSPSFQGQVTISADKSISTAYLQWSSLKASDTAMYYC</td>\n",
       "      <td>ARHRRDYYDSSGYYYGGYYYYGMDV</td>\n",
       "      <td>ARHRRDYYDSSGYYYGGYYYYGMDV</td>\n",
       "      <td>WGQGTTVTVSS</td>\n",
       "      <td>WGQGTTVTVSS</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>20.0</td>\n",
       "      <td>-0.456397</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>generated_sequence_6</td>\n",
       "      <td>EVQLVQSGAEVKKPGESLRISCKGSGYSFSTYWISWVRQMPGKGLE...</td>\n",
       "      <td>0</td>\n",
       "      <td>human</td>\n",
       "      <td>H</td>\n",
       "      <td>0.0</td>\n",
       "      <td>180.750</td>\n",
       "      <td>0</td>\n",
       "      <td>124</td>\n",
       "      <td>human</td>\n",
       "      <td>...</td>\n",
       "      <td>NYSPSFQ-GQVTISADKSISTAYLQWSSLKASDTAMYYC</td>\n",
       "      <td>NYSPSFQGQVTISADKSISTAYLQWSSLKASDTAMYYC</td>\n",
       "      <td>ARRGGYCSSTSCYDAFDI</td>\n",
       "      <td>ARRGGYCSSTSCYDAFDI</td>\n",
       "      <td>WGQGTMVTVSS</td>\n",
       "      <td>WGQGTMVTVSS</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>13.0</td>\n",
       "      <td>-0.412610</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>7 rows × 40 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                     Id                                           sequence  \\\n",
       "0           wt_sequence  EVQLVQSGAEVKKPGESLRISCKGSGYSFSTYWISWVRQMPGKGLE...   \n",
       "1  generated_sequence_1  EVQLVQSGAEVKKPGESLRISCKGSGYSFSTYWISWVRQMPGKGLE...   \n",
       "2  generated_sequence_2  EVQLVQSGAEVKKPGESLRISCKGSGYSFSTYWISWVRQMPGKGLE...   \n",
       "3  generated_sequence_3  EVQLVQSGAEVKKPGESLRISCKGSGYSFSTYWISWVRQMPGKGLE...   \n",
       "4  generated_sequence_4  EVQLVQSGAEVKKPGESLRISCKGSGYSFSTYWISWVRQMPGKGLE...   \n",
       "5  generated_sequence_5  EVQLVQSGAEVKKPGESLRISCKGSGYSFSTYWISWVRQMPGKGLE...   \n",
       "6  generated_sequence_6  EVQLVQSGAEVKKPGESLRISCKGSGYSFSTYWISWVRQMPGKGLE...   \n",
       "\n",
       "   domain_no hmm_species chain_type  e-value    score  seqstart_index  \\\n",
       "0          0       human          H      0.0  185.000               0   \n",
       "1          0       human          H      0.0  183.375               0   \n",
       "2          0       human          H      0.0  169.500               0   \n",
       "3          0       human          H      0.0  183.250               0   \n",
       "4          0       human          H      0.0  182.500               0   \n",
       "5          0       human          H      0.0  167.750               0   \n",
       "6          0       human          H      0.0  180.750               0   \n",
       "\n",
       "   seqend_index identity_species  ...  \\\n",
       "0           115            human  ...   \n",
       "1           117            human  ...   \n",
       "2           131            human  ...   \n",
       "3           123            human  ...   \n",
       "4           122            human  ...   \n",
       "5           131            human  ...   \n",
       "6           124            human  ...   \n",
       "\n",
       "                              fwr3_aa_gaps  \\\n",
       "0  NYSPSFQ-GQVTISADKSISTAYLQWSSLKASDTAMYYC   \n",
       "1  NYSPSFQ-GQVTISADKSISTAYLQWSSLKASDTAMYYC   \n",
       "2  NYSPSFQ-GQVTISADKSISTAYLQWSSLKASDTAMYYC   \n",
       "3  NYSPSFQ-GQVTISADKSISTAYLQWSSLKASDTAMYYC   \n",
       "4  NYSPSFQ-GQVTISADKSISTAYLQWSSLKASDTAMYYC   \n",
       "5  NYSPSFQ-GQVTISADKSISTAYLQWSSLKASDTAMYYC   \n",
       "6  NYSPSFQ-GQVTISADKSISTAYLQWSSLKASDTAMYYC   \n",
       "\n",
       "                          fwr3_aa_no_gaps               cdr3_aa_gaps  \\\n",
       "0  NYSPSFQGQVTISADKSISTAYLQWSSLKASDTAMYYC              ARGYG----IFDY   \n",
       "1  NYSPSFQGQVTISADKSISTAYLQWSSLKASDTAMYYC              ARRGYS--YGYDY   \n",
       "2  NYSPSFQGQVTISADKSISTAYLQWSSLKASDTAMYYC  ARHSPRGGYYDSSGYYPYYYYGMDV   \n",
       "3  NYSPSFQGQVTISADKSISTAYLQWSSLKASDTAMYYC          ARHGDYYDSSGYYPFDY   \n",
       "4  NYSPSFQGQVTISADKSISTAYLQWSSLKASDTAMYYC           ARRGAGGTYYYYGMDV   \n",
       "5  NYSPSFQGQVTISADKSISTAYLQWSSLKASDTAMYYC  ARHRRDYYDSSGYYYGGYYYYGMDV   \n",
       "6  NYSPSFQGQVTISADKSISTAYLQWSSLKASDTAMYYC         ARRGGYCSSTSCYDAFDI   \n",
       "\n",
       "             cdr3_aa_no_gaps fwr4_aa_gaps fwr4_aa_no_gaps leader follow  \\\n",
       "0                  ARGYGIFDY  WGQGTLVTVSS     WGQGTLVTVSS    NaN    NaN   \n",
       "1                ARRGYSYGYDY  WGQGTLVTVSS     WGQGTLVTVSS    NaN    NaN   \n",
       "2  ARHSPRGGYYDSSGYYPYYYYGMDV  WGQGTTVTVSS     WGQGTTVTVSS    NaN    NaN   \n",
       "3          ARHGDYYDSSGYYPFDY  WGQGTLVTVSS     WGQGTLVTVSS    NaN    NaN   \n",
       "4           ARRGAGGTYYYYGMDV  WGQGTTVTVSS     WGQGTTVTVSS    NaN    NaN   \n",
       "5  ARHRRDYYDSSGYYYGGYYYYGMDV  WGQGTTVTVSS     WGQGTTVTVSS    NaN    NaN   \n",
       "6         ARRGGYCSSTSCYDAFDI  WGQGTMVTVSS     WGQGTMVTVSS    NaN    NaN   \n",
       "\n",
       "  Levenshtein_distance   ll_mean  \n",
       "0                  NaN       NaN  \n",
       "1                  5.0 -0.411147  \n",
       "2                 20.0 -0.479769  \n",
       "3                  9.0 -0.419964  \n",
       "4                 12.0 -0.452040  \n",
       "5                 20.0 -0.456397  \n",
       "6                 13.0 -0.412610  \n",
       "\n",
       "[7 rows x 40 columns]"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Load the CSV file into a DataFrame\n",
    "df = pd.read_csv('ProGen2_OAS_sequences.csv')\n",
    "\n",
    "# Display the first few rows of the DataFrame\n",
    "df.head(10)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6a51e194-cd11-4f5b-82bd-d5e05b6ab8f0",
   "metadata": {},
   "source": [
    "The DataFrame above contains the Wild-type sequence as well as the generated sequences. Information on species, V and J gene segments and identities, frameworks and CDR3 region sequences, Levenshtein distance and ll_mean values are all given. \n",
    "\n",
    "Sequences were filterd and ranked based on ll_mean. In the context of ProGen2-OAS, which has been specifically trained on a diverse set of antibody sequences from the Observed Antibody Space (OAS) database, a lower ll_mean value has many implications. The model has learned the patterns associated with functional antibodies, so a lower log-likelihood mean indicates that the generated sequence aligns well with these functional patterns. Furthermore, since the training data has been clustered to reduce redundancy and ensure diversity, a lower ll_mean suggests that the sequence is a good match to the natural variation observed in the antibody space. \n",
    "\n",
    "The goal is to provide easy and programmatic access to language models like ProGen2-OAS, via the BioLM API, to optimize\n",
    "antibody properties including thermal stability, expression, aggregation propensity, and solubility. "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3c7aa5b9-4f74-47a9-9719-672c1ca2c688",
   "metadata": {},
   "source": [
    "# NanoBodyBuilder 2:\n",
    "\n",
    "NanoBodyBuilder2 can be used to produce models of nanobody. If we only have the heavy chain sequence of a conventional antibody (from humans, mice, etc.), it does not automatically become a nanobody. However, for this use-case, we can use NanoBodyBuilder2 as a bioinformatics tool to predict the structure of our ProGen2-OAS generated sequences of Utomilumab.  To learn more about NanoBodyBuilder2, click <u>[here](https://github.com/oxpig/ANARCI/issues/12)<u><i>\n",
    "\n",
    "<i>NanoBodyBuilder 2 requires openmm, pdbfixer, anarci and ImmuneBuilder.\n",
    "Information on how to install and run these packages can be found <u>[here](https://opig.stats.ox.ac.uk/webapps/sabdab-sabpred/sabpred/nanobodybuilder2/)<u>\n",
    "\n",
    "<i>Further troubleshooting information can be found <u>[here](https://github.com/oxpig/ANARCI/issues/12)<u><i>\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "d364f299-9101-4037-a331-9a9d51340a3d",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Import necessary packages and dependencies \n",
    "\n",
    "import pandas as pd\n",
    "from ImmuneBuilder import NanoBodyBuilder2\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "a43b0152-7432-42f1-9caa-af33d3f38ff7",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Saved the generated nanobody structure to wt_sequence.pdb\n",
      "Saved the generated nanobody structure to generated_sequence_1.pdb\n",
      "Saved the generated nanobody structure to generated_sequence_2.pdb\n",
      "Saved the generated nanobody structure to generated_sequence_3.pdb\n",
      "Saved the generated nanobody structure to generated_sequence_4.pdb\n",
      "Saved the generated nanobody structure to generated_sequence_5.pdb\n",
      "Saved the generated nanobody structure to generated_sequence_6.pdb\n"
     ]
    }
   ],
   "source": [
    "# Load the CSV file generated in Function 2 above into a DataFrame: \"ProGen2_OAS_sequences.csv\"\n",
    "\n",
    "csv_file_path = 'ProGen2_OAS_sequences.csv' #CSV file generated in Function 2 above.\n",
    "sequences_df = pd.read_csv(csv_file_path)\n",
    "\n",
    "# Initializing NanoBodyBuilder2\n",
    "predictor = NanoBodyBuilder2()\n",
    "\n",
    "# Looping through each row in the DataFrame from above\n",
    "\n",
    "for index, row in sequences_df.iterrows():    # iterrows() is a generator that iterates over the rows of the DataFrame.\n",
    "    sequence_id = row['Id']                   # Accessing the value in the column named 'Id'.\n",
    "    sequence = {'H': row['sequence']}         # Accessing the value in the column named 'sequence'. \n",
    "\n",
    "                                              \n",
    "    nanobody = predictor.predict(sequence)    # Generating the nanobody/antibody structure.\n",
    "    output_file = f\"{sequence_id}.pdb\"        # Defining the output file path using the sequence ID.\n",
    "    nanobody.save(output_file)                # Saving the nanobody structure to a PDB file.\n",
    "\n",
    "    print(f\"Saved the generated nanobody structure to {output_file}\") # We now have a PDB files of the Utomilumab generated sequences. \n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "9d97f5cf-b277-4188-b723-41a8b9ede923",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "REMARK  NANOBODY STRUCTURE MODELLED USING NANOBODYBUILDER2                      \n",
      "REMARK  STRUCTURE REFINED USING OPENMM 8.1.1, 2024-03-29                        \n",
      "ATOM      1  N   GLU H   1      -8.448  -4.352 -18.972  1.00  1.28           N  \n",
      "ATOM      2  H   GLU H   1      -7.935  -3.488 -18.852  1.00  1.28           H  \n",
      "ATOM      3  H2  GLU H   1      -8.232  -4.753 -19.874  1.00  1.28           H  \n",
      "ATOM      4  H3  GLU H   1      -9.433  -4.143 -18.892  1.00  1.28           H  \n",
      "ATOM      5  CA  GLU H   1      -8.045  -5.278 -17.891  1.00  1.28           C  \n",
      "ATOM      6  HA  GLU H   1      -8.618  -6.202 -17.973  1.00  1.28           H  \n",
      "ATOM      7  C   GLU H   1      -8.356  -4.649 -16.531  1.00  1.28           C  \n",
      "ATOM      8  CB  GLU H   1      -6.564  -5.621 -18.071  1.00  1.28           C  \n"
     ]
    }
   ],
   "source": [
    "#Lets view the first few lines of the pdb files:\n",
    "\n",
    "# Specify the path to your PDB file.\n",
    "pdb_file_path = 'generated_sequence_1.pdb'\n",
    "\n",
    "# Number of lines to display from the beginning of the file.\n",
    "num_lines = 10\n",
    "\n",
    "# Open the file and read the specified number of lines.\n",
    "with open(pdb_file_path, 'r') as file:\n",
    "    for _ in range(num_lines):\n",
    "        line = file.readline()\n",
    "        print(line, end='')  # Print each line, end='' prevents adding an extra newline.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fd393f49-2300-455f-a801-4fd14940c590",
   "metadata": {},
   "source": [
    "\n",
    "\n",
    "We have demonstrated that by leveraging the BioLM.ai API and ProGen2-OAS model, researchers can rapidly generate and structurally predict a diverse library of Utomilumab antibody sequences with new CDR3 regions.\n",
    "\n",
    "CDR3 regions are the most variable parts of the antibody and play a pivotal role in determining the specificity of the antigen binding site. Due to their immense sequence diversity, remarkable conformational flexibility that changes upon antigen binding, and limited available structural data compared to other antibody regions, modeling and predicting the structure of CDR3 loops remains a significant challenge for antibody scientists. \n",
    "\n",
    "This tutorial equips experts with the tools to navigate the exciting world of antibody engineering. We eagerly anticipate experts utilizing our platform to create novel antibodies with customized sequences and CDR3 regions, advancing the field through their innovative research."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d0d59b38-55a0-476c-82e5-29446a53b110",
   "metadata": {},
   "source": [
    "## Next Steps\n",
    "\n",
    "Check out additional tutorials at <u>[jupyter.biolm.ai](https://jupyter.biolm.ai)<u>\n",
    "\n",
    "To explore additional models and functionality, head over to our <u>[BioLM Documentation]<u>(https://docs.biolm.ai)<u>\n",
    "\n",
    "   \n",
    "#### See more use-cases and APIs on your <u>[BioLM Console Catalog](https://biolm.ai/console/catalog/).<u>\n",
    "<br>\n",
    "\n",
    "##### BioLM hosts deep learning models and runs inference at scale. You do the science.\n",
    "<br>\n",
    "\n",
    "<table class=\"jupyter-biolm-header-table\" style=\"width: 100%; border-collapse: collapse; background-color: white; float: left;\">\n",
    "    <tr>\n",
    "        <td style=\"text-align: left; vertical-align: middle; background-color: white;\">\n",
    "            <img src=\"https://d31e6ufxekikrt.cloudfront.net/static/ui/images/console-overview/enzyme_engineering_icon.png\"  style=\"height: 40px; float: left; padding-right: 10px;\"> Enzyme Engineering\n",
    "        </td>\n",
    "        <td style=\"text-align: left; vertical-align: middle; background-color: white;\">\n",
    "            <img src=\"https://d31e6ufxekikrt.cloudfront.net/static/ui/images/console-overview/antibody_engineering_icon.png\"  style=\"height: 40px; float: left; padding-right: 10px;\"> Antibody Engineering\n",
    "        </td>\n",
    "        <td style=\"text-align: left; vertical-align: middle; background-color: white;\">\n",
    "            <img src=\"https://d31e6ufxekikrt.cloudfront.net/static/ui/images/console-overview/biosecurity_icon.png\"  style=\"height: 40px; float: left; padding-right: 10px;\"> Biosecurity\n",
    "        </td>\n",
    "    </tr>\n",
    "    <tr>\n",
    "        <td style=\"text-align: left; vertical-align: middle; background-color: white;\">\n",
    "            <img src=\"https://d31e6ufxekikrt.cloudfront.net/static/ui/images/console-overview/single_cell_genomics_icon.png\"  style=\"height: 40px; float: left; padding-right: 10px;\"> Single-Cell Genomics\n",
    "        </td>\n",
    "        <td style=\"text-align: left; vertical-align: middle; background-color: white;\">\n",
    "            <img src=\"https://d31e6ufxekikrt.cloudfront.net/static/ui/images/console-overview/dna_seq_modeling_icon.png\"  style=\"height: 40px; float: left; padding-right: 10px;\"> DNA Sequence Modelling\n",
    "        </td>\n",
    "        <td style=\"text-align: left; vertical-align: middle; background-color: white;\">\n",
    "            <img src=\"https://d31e6ufxekikrt.cloudfront.net/static/ui/images/console-overview/finetuning_icon.png\"  style=\"height: 40px; float: left; padding-right: 10px;\"> Finetuning\n",
    "        </td>\n",
    "    </tr>\n",
    "</table>\n",
    "\n",
    "#### [**Contact us**](https://biolm.ai/ui/contact-us/) to learn more.\n",
    "    \n",
    "    \n",
    "Author: Zeeshan Siddiqui,   Bioinformaitican @BioLM.ai\n",
    "\n",
    "Email ID: zsiddiqui@biolm.ai\n",
    "\n"
   ]
  }
 ],
 "metadata": {
  "environment": {
   "kernel": "conda-root-py",
   "name": "workbench-notebooks.m115",
   "type": "gcloud",
   "uri": "gcr.io/deeplearning-platform-release/workbench-notebooks:m115"
  },
  "kernelspec": {
   "display_name": "Python 3 (ipykernel) (Local)",
   "language": "python",
   "name": "conda-root-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
